### DISCUSSION TITLE
How to properly set precision of whole numbers and decimals?

### DISCUSSION THREAD

**Discussion by Andre601** [ezylang/EvalEx#285] [Q&A] [answered]
I use EvalEx in one of my projects and some people report, that the returned number seems to not be accurate, as it rounds certain digits to 0s.

As an example, just providing 1234567891 results in 1234568000 being returned which shouldn't be the case.

Right now am I using the following:
```java
Expression exp = new Expression(expression);
return exp.eval().setScale(3, RoundingMode.HALF_UP);
```

I assume to fix this would I use `setPrecision` before the `eval()`?
```java
Expression exp = new Expression(expression);
return exp.setPrecision(10).eval().setScale(3, RoundingMode.HALF_UP);
```

And if the above is indeed the case, would there be any way to make EvalEx only round the whole number, if A) the scale is set to zero (no decimals) and/or B) the decimals result in rounding the next whole number up?

Example:
A number `123456789.1` would become `123456789` while `123456789.9` would become `123456790`.

**Comment by RobertZenz** [comment, marked as answer]
No, precision/scale of a [BigDecimal](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html) works by setting the significant number of digits, not decimal places. So if you have the value "123456798" and you set the precision to "3" you get "12300000". Same goes if you have "1.23456789" which becomes "1.23".

What you seem to want to do is to set a "high enough" precision to do your calculation, and then *[round](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#round-java.math.MathContext-)* the result.

**Reply by Andre601** [reply]
Thanks for the reply.

So if I get you correctly could I just set a high number (I.e. `new MathContext(128)` for 128 digits with rounding mode `HALF_UP`) for the precision, make EvalEx calculate everything as normal and then apply the `round(MathContext)` to the final BigDecimal to round?
In such a case, I assume I would then provide the `new MathContext(precision, roundingMode)` to allow people to set their own precision and rounding behaviours?

**Reply by RobertZenz** [reply]
Yes.

**Reply by Andre601** [reply]
Okay, I'm encountering yet another issue with this entire precision and rounding stuff and would appreciate some more help with solving it.

Basically, users report that the number they set isn't just counting decimal points but the whole number itself.
As an example, with the value set to 1 doe the following results return:
```
- 22.5    -> 20
- 13150.5 -> 10000
- 0.5     -> 0.5
- 0.55    -> 0.6
```
...while the numbers wanted are...
```
- 22.5    -> 23
- 13150.5 -> 13151
- 0.5     -> 1
- 0.55    -> 1
```

I assume this is to blame on how BigDecimal (And/or MathContext) is handling the provided values to be rounded and shortened, but I would appreciate if you could share any possible workaround for this inconvenience here.

The current code I use:
```java
    private String evaluate(String expression, int decimals, RoundingMode roundingMode){
        try{
            return new Expression(expression)
                .setPrecision(128) // Just set to a high number to use.
                .eval()
                .round(new MathContext(decimals, roundingMode))
                .toPlainString();
        }catch(Exception ex){
            ex.printStackTrace();
            return null;
        }
    }
```

**Reply by RobertZenz** [reply]
> Basically, users report that the number they set isn't just counting decimal points but the whole number itself.

Yes, that's what I said...

 > No, precision/scale of a [BigDecimal](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html) works by setting the significant number of digits, not decimal places.

**Reply by RobertZenz** [reply]
> ...while the numbers wanted are...
 > 
 >     - 22.5    -> 23
 >     - 13150.5 -> 13151
 >     - 0.5     -> 1
 >     - 0.55    -> 1

That means that you want to round to zero decimal places, but do the calculation with a high enough precision.

**Reply by Andre601** [reply]
This doesn't answer the question I have here, which is how to set the decimal points to use...
After all can I not set the precision to the size of the final number, since I don't have that final number yet available...

Like the way you explain this - and how I understand this - Would I need to

- Evaluate the number
- Get the length up until the dot (i.e. `123.45` would be a length of 3)
- Add the amount of decimals the user want to include to the number length.
- Evaluate the number again, but this time use the number acquired before to set the max length
- Apply MathContext to this stuff
- get String.

???

Like as much as I try to understand this is your reply here not that helpful as it keeps me still confused on how to properly do this...

**Reply by RobertZenz** [reply]
Now I see where the confusion is coming from, my mistake for linking against the wrong method, what you're looking for is the [`setScale` method of `BigDecimal`](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#setScale-int-int-).

**Reply by Andre601** [reply]
> Now I see where the confusion is coming from, my mistake for linking against the wrong method, what you're looking for is the [`setScale` method of `BigDecimal`](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#setScale-int-int-).

So, just to make things clear here:
If I use `setScale(0, RoundingMode.HALF_UP)`, it would give me the numbers I mentioned [here](https://github.com/uklimaschewski/EvalEx/discussions/285#discussioncomment-2986137)?

**Reply by RobertZenz** [reply]
Yes.
